jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/util/DiscreteDistribution.java

```diff
@@ -165,43 +164,42 @@ public static void normalize(double[] counts, double alpha) {
    * @see #mean(double[][])
    * @param distributions the distributions whose mean is to be calculated
    * @return the mean of the distributions
+   * @throws IllegalArgumentException if the outer distributions array is empty, or if any of the
+   *     inner distribution arrays are empty, or if not all of the inner distribution arrays are the
+   *     same length
    */
   public static double[] mean(Collection<double[]> distributions) {
-    Preconditions.checkArgument(
-        !distributions.isEmpty(), "Distribution collection must be non-empty");
-    // TODO: Consider checking that the inner arrays of `distributions` are non-empty
-    // TODO: Consider using com.google.common.math.Stats.meanOf
-    Iterator<double[]> iter = distributions.iterator();
-    double[] first = iter.next();
-    double[][] dArray = new double[distributions.size()][first.length];
-    dArray[0] = first;
-    for (int i = 1; i < dArray.length; i++) {
-      dArray[i] = iter.next();
-    }
+    checkArgument(!distributions.isEmpty(), "Distribution collection must be non-empty");
+    checkArraysAreNonEmpty(distributions);
+    checkArraysAreSameLength(distributions);
+
+    return distributions.stream().mapToDouble(Stats::meanOf).toArray();
+  }
 
-    return mean(dArray);
+  private static void checkArraysAreNonEmpty(Collection<double[]> arrays) {
```

Comment:
I would be inclined to do these checks (`checkArraysAre*()`) inline since we only ever use them together, and once.

It also makes me a little bit sad that we're iterating over the input 3x to do those checks rather than doing something like this:

```
int length = Iterables.get(arrays, 0).length;
double[] dMean = new double[length];
int i = 0;
for (double[] distribution : distributions) {
  checkArgument(distribution.length != 0, "all distributions must be non-empty");
  checkArgument(distribution.length == length, "all distributions must have the same number of elements");
  dMean[i++] = Stats.meanOf(distribution);
}
```

but I suppose that part of that reaction is that I haven't completely internalized Stream yet.

Reply:
You make a rather good point, @jrtom, about the 3x iteration. I've refactored the code in the latest commit to something similar to your suggestion above, but using a single `Stream` instead of a foreach loop.

WDYT of the code now?